src/Suave/Logging/LogLine.fs

```diff
@@ -13,23 +15,45 @@ type LogLine =
     level         : LogLevel
     /// the source of the log line, e.g. 'ModuleName.FunctionName'
     path          : string
-    /// the message that the application wants to log
-    message       : string
+    /// the template message that the application wants to log
+    template      : string
+    /// a function which can format the log template message. When the
+    /// message is pre-formatted, this is None. When the log message 
+    /// contains structured data, this function can be used to render
+    /// the message template with the data.
+    formatter     : LogTemplateFormatter option
+    /// any key-value based data to log
+    data          : Map<string, obj>
     /// an optional exception
     ``exception`` : exn option
     /// timestamp when this log line was created
     tsUTCTicks    : int64 }
-    
+
+    /// Formats the log line as a string
+    member x.formatMessage () =
+        match x.formatter with
+        | None -> x.template
+        | Some fmter -> fmter x.template x.GetDataByName
+
+    /// Gets a data value
+    member x.GetDataByName name = Option.toObj (x.data.TryFind name)
+
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module LogLine =
   let mk path level trace ex message =
-    { message       = message
+    { template      = message
+      formatter     = None
+      data          = Map.empty
       level         = level
       path          = path
       ``exception`` = ex
       trace         = trace
       tsUTCTicks  = Globals.utcNow().Ticks }
 
+  /// assists with type inference, avoiding ugly 'box' calls for each
+  /// value item when constructing the log data map 
+  let mapLogData (data: (string*obj) list) = Map.ofList data
```

Comment:
1. Moving complexity around? Not really. You can still have an overload that takes an anonymous object and reflects on it. This way though, you get speed in the default case, speed if you implement ToValue or slow reflection in any other case.
 2. `{???}` is equivalent to `{startData}` because `startData` is positional, just like ???
 3. In the proposed model, inc Logary, the log message creator uses the API to add Fields that contain values, but you can invoke Message.fieldWithUnit from `logger.Information` just as well, so you don't need the ugly `convertToStructure`, because that's implicit in `Message.fieldWithUnit` since it sets a Field-value in the Message structure. The cognitive dissonance you're experiencing is that you don't have anonymous objects in F#, but you're still thinking in terms of them. But yes, if you have records, you'd either need to provide a ToValue method on it OR you can let it default to reflection, using the existing infrastructure that you've written. I suggest you write it this way.
 4. The lens doesn't add complexity, imo. It adds structure and simplicity as it separates how you change Value values from how they are structured, at the point of the call-site. And this is all about complexity *at the call site*, right?

In the proposed model I left out the parsing, since I figured we'd use your work for it; your parsing was good. The storage for the template is [the message string](https://github.com/logary/logary/blob/695ca9df777503bb9a477e4136bce25dbc84878b/src/Logary/DataModel.fs#L478); compared to your work, we delay the parsing until we need it – which we do at the point when we output the value into a Target (Sink). That way you can use the hash of the template to create a histogram of logged messages, without the templated values changing the uniqueness.

Does this make sense?